designs/0005-static-matrices.md

```diff
@@ -0,0 +1,252 @@
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the `varmat` entries to zero.
+
+The assumptions that makes this work is after a subset of a `varmat` is overwritten,
+that part of the `varmat` cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a `varmat` variable would make it so that views
+point into that part of the `varmat` variable's memory no longer point to the expected
+variables.
+
+# Stanc3 implementation
+[stanc3-implementation]: #stanc3-implementation
+
+The language implementation of `varmat` is yet undefined. It was left unresolved because
+enough was clear from the initial discussions of how to handle the Math implementation
+regardless of how the language was handled.
+
+At this point, the Math maturing to the point that it is possible to compile a model using
+`varmat`s values with a development version of the Stanc3 compiler (available
+[here](https://github.com/stan-dev/stanc3/pull/755), though it may take some effort to run since
+it is a work in progress). This Stanc3 compiler defines any variable with a name ending in
+`_`varmat`` to use a `varmat` type.
+
+A basic bernoulli MRP model with five population level covariates and seven group terms is available
+[here](https://github.com/bbbales2/computations_in_glms/blob/master/benchmark_model/mrp_`varmat`.stan).
+
+The initial `varmat` implementation of this model achieves a per-gradient speedup of around two.
+
+Now that `varmat` variables can be assigned, it is possible that the compiler silently replace every
+`matvar` in Stan with a `varmat`, and none of the standard Stan use cases would be affected.
+
+There are four major implementation strategies available at the Stanc3 level.
+
+1. Implement a new Stan type to differentiate explicitly between `matvar` and `varmat` types
+
+2. Use `matvar` types for any function in the Stan program that does not support `varmat`
+
+3. Use `varmat` types by default, but if variable assignment is detected, or a `varmat` is used
+in an unsupported function convert to and from `matvar` automatically.
+
+4. Use `varmat` types everywhere
+
+The original proposal was leaning towards #1 because it most naturally corresponds to the Math
+implementation. However, this leaves the problem of explicitly declaring `varmat` to the user. To
+avoid duplicating the user's manual and function reference for `varmat` and `matvar` functions, there
+would also need to be some automatic, under the hood conversions between the types. This automatic
+converting would make it so that if a `varmat` is forgotten anywhere, the entire program will be
+slowed down. This sort of explicit typing shouldn't be too much of a problem in development models
+or generated Stan code, but in large, user-written models this would be error prone to implement
+and maintain.
```

Comment:
1 also duplicates the list of signatures (longer error messages). Though that can be assumed from the duplicating function reference I guess.